library/OpenTelemetry/AWSXRay/TraceInfo.hs

```diff
@@ -0,0 +1,120 @@
+{-# LANGUAGE NamedFieldPuns #-}
+
+-- |
+--
+-- <https://docs.aws.amazon.com/xray/latest/devguide/xray-concepts.html#xray-concepts-tracingheader>
+--
+-- Example:
+--
+-- @Root=1-5759e988-bd862e3fe1be46a994272793;Parent=53995c3f42cd8ad8;Sampled=1@
+--
+-- @Root=1-{epoch}-{unique}[;Parent={spanId}][;Sampled={1|0}][;meta=attr...]@
+--
+module OpenTelemetry.AWSXRay.TraceInfo
+  ( TraceInfo(..)
+  , fromXRayHeader
+  , toXRayHeader
+  ) where
+
+import Prelude
+
+import Control.Error.Util (note)
+import Data.Bifunctor (first)
+import Data.ByteString (ByteString)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Char8 as BS8
+import qualified OpenTelemetry.AWSXRay.Baggage as Baggage
+import OpenTelemetry.Baggage (Baggage)
+import OpenTelemetry.Trace (SpanContext(..))
+import OpenTelemetry.Trace.Core
+  (defaultTraceFlags, isSampled, setSampled, unsetSampled)
+import OpenTelemetry.Trace.Id
+  ( Base(..)
+  , baseEncodedToSpanId
+  , baseEncodedToTraceId
+  , spanIdBaseEncodedByteString
+  , traceIdBaseEncodedByteString
+  )
+import qualified OpenTelemetry.Trace.TraceState as TS
+
+-- | The data to read/write from the @X-Amzn-TraceId@ header
+data TraceInfo = TraceInfo
+  { spanContext :: SpanContext
+  , baggage :: Maybe Baggage
+  }
+  deriving stock Show
+
+fromXRayHeader :: ByteString -> Either String TraceInfo
+fromXRayHeader bs = do
+  kv <- bsToKeyValues bs
+
+  root <- note "Root not present" $ lookup "Root" kv
+  traceId <- case bsSplitOn '-' root of
+    ["1", epoch, unique] -> do
+      -- epoch can be "at most" 8 hex and we should zero-pad the
+      -- short values. NB. This relies on hex being 1 char per char
```

Comment:
:thinking:

Dumb question warning: why would it not be "1 char per char"?

Reply:
I guess it's just that it relies on it being _hex_. A `ByteString` in some other base or encoding would behave differently when you do a `Char`-based operation, I think. I'd _guess_ that `'0'` is one of those characters this is going to work no matter the base or encoding, but I'm not sure! I can definitely nuke the comment if it's more confusing than helpful.